core/server/src/test/java/alluxio/worker/file/FileDataManagerTest.java

```diff
@@ -126,6 +140,76 @@ public void clearPersistedFilesTest() {
   }
 
   /**
+   * Tests the rate limiting functionality for asynchronous persistence.
+   */
+  @Test
+  public void persistFileRateLimitingTest() throws Exception {
+    long fileId = 1;
+    List<Long> blockIds = Lists.newArrayList(1L, 2L, 3L);
+
+    // mock block worker
+    BlockWorker blockWorker = Mockito.mock(BlockWorker.class);
+    FileInfo fileInfo = new FileInfo();
+    fileInfo.setPath("test");
+    Mockito.when(blockWorker.getFileInfo(fileId)).thenReturn(fileInfo);
+    BlockReader reader = Mockito.mock(BlockReader.class);
+    for (long blockId : blockIds) {
+      Mockito.when(blockWorker.lockBlock(Sessions.CHECKPOINT_SESSION_ID, blockId))
+          .thenReturn(blockId);
+      Mockito
+          .when(blockWorker.readBlockRemote(Sessions.CHECKPOINT_SESSION_ID, blockId, blockId))
+          .thenReturn(reader);
+      BlockMeta mockedBlockMeta = PowerMockito.mock(BlockMeta.class);
+      Mockito.when(mockedBlockMeta.getBlockSize()).thenReturn(100L);
+      Mockito
+          .when(blockWorker.getBlockMeta(Sessions.CHECKPOINT_SESSION_ID, blockId, blockId))
+          .thenReturn(mockedBlockMeta);
+    }
+
+    Configuration conf = WorkerContext.getConf();
+    conf.set(Constants.WORKER_FILE_PERSIST_RATE_LIMIT_ENABLED, "true");
+    conf.set(Constants.WORKER_FILE_PERSIST_RATE_LIMIT, "100");
+
+    FileDataManager manager = new FileDataManager(blockWorker);
+
+    // mock ufs
+    UnderFileSystem ufs = Mockito.mock(UnderFileSystem.class);
+    String ufsRoot = new Configuration().get(Constants.UNDERFS_ADDRESS);
+    Mockito.when(ufs.exists(ufsRoot)).thenReturn(true);
+    Whitebox.setInternalState(manager, "mUfs", ufs);
+
+    // Setup a mock rate limiter.
+    final MockRateLimiter mockRateLimiter = new MockRateLimiter(
+        WorkerContext.getConf().getBytes(Constants.WORKER_FILE_PERSIST_RATE_LIMIT));
+    Whitebox.setInternalState(
+        manager, "mPersistenceRateLimiter", mockRateLimiter.getGuavaRateLimiter());
+
+    OutputStream outputStream = Mockito.mock(OutputStream.class);
+
+    // mock BufferUtils
+    PowerMockito.mockStatic(BufferUtils.class);
+
+    String dstPath = PathUtils.concatPath(ufsRoot, fileInfo.getPath());
+    Mockito.when(ufs.create(dstPath)).thenReturn(outputStream);
+
+    manager.lockBlocks(fileId, blockIds);
+    manager.persistFile(fileId, blockIds);
+
+    List<String> expectedEvents = Lists.newArrayList("R0.00", "R1.00", "R1.00");
+    assertEquals(expectedEvents, mockRateLimiter.readEventsAndClear());
+
+    // Simulate waiting for 1 second.
+    mockRateLimiter.sleepMillis(1000);
+
+    manager.lockBlocks(fileId, blockIds);
+    manager.persistFile(fileId, blockIds);
+
+    // The first write will go through immediately without throttling.
```

Comment:
Could you also repeat the test of persisting twice but with sleeping? Then we should expect:
`"R0.00", "R1.00", "R1.00"` and `"R1.00", "R1.00", "R1.00"`